content/00.build/65.developer-reference/50.era-contracts/10.l1-contracts.md

```diff
@@ -17,108 +17,202 @@ items: [{
+<!-- TODO: update ecosystem contracts link -->
+
+## Diamond (also mentioned as State Transition contract)
+
+Technically, this L1 smart contract acts as a connector between Ethereum (L1) and hyperchain (L2).
+It checks the validity proof and data availability,
+handles L2 <-> L1 communication, finalizes L2 state transition, and more.
+
+There are also important contracts deployed on the L2 that can also execute logic called *system contracts*.
+Using L2 <-> L1 communication can affect both the L1 and the L2.
 
-Each of the facets has an associated parameter that indicates if it is possible to freeze access to the facet.
+![diamondProxy.png](/images/developer-reference/l1-smart-contracts/Diamond-scheme.png)
 
-Privileged actors can freeze the **diamond** (not a specific facet!) and all facets
-with the marker `isFreezable` should be inaccessible until the governor unfreezes the diamond.
+### DiamondProxy
 
-## Diamond
+The main contract uses [EIP-2535](https://eips.ethereum.org/EIPS/eip-2535) diamond proxy pattern. It is an in-house
+implementation that is inspired by the [mudgen reference implementation](https://github.com/mudgen/Diamond). It has no
+external functions, only the fallback that delegates a call to one of the facets (target/implementation contract). So
+even an upgrade system is a separate facet that can be replaced.
+
+One of the differences from the reference implementation is access freeze-ability. Each of the facets has an associated
+parameter that indicates if it is possible to freeze access to the facet. Privileged actors can freeze the **diamond**
+(not a specific facet!) and all facets with the marker `isFreezable` should be inaccessible until the governor or admin
+unfreezes the diamond. Note that it is a very dangerous thing since the diamond proxy can freeze the upgrade system and then
+the diamond will be frozen forever.
+
+The diamond proxy pattern is very flexible and extendable.
+For now, it allows splitting implementation contracts by their logical meaning,
+removes the limit of bytecode size per contract and implements security features such as freezing.
+In the future, it can also be viewed as [EIP-6900](https://eips.ethereum.org/EIPS/eip-6900)
+for [zkStack](https://blog.matter-labs.io/introducing-the-zk-stack-c24240c2532a),
+where each hyperchain can implement a sub-set of allowed implementation contracts.
 
-Technically, this L1 smart contract acts as a connector between Ethereum (L1) and ZKsync (L2).
-This contract checks the validity proof and data availability, handles
-L2 <-> L1 communication, finalizes L2 state transition, and more.
+### GettersFacet
+
+Separate facet, whose only function is providing `view` and `pure` methods. It also implements
+[diamond loupe](https://eips.ethereum.org/EIPS/eip-2535#diamond-loupe) which makes managing facets easier.
```

Comment:
nit: The specific text `diamond-loupe` does not exist in the document. Perhaps we should use the following direct link instead? https://eips.ethereum.org/EIPS/eip-2535#inspecting-facets--functions

Reply:
Good catch, yes the link should be fixed